### PR TITLE
[SAIC-673] Fix images filtering

### PR DESCRIPTION
**Pull request by Settis**

### CODE REVIEW

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module tests.cloudferrylib.os.image.test_image_filters
tests/cloudferrylib/os/image/test_image_filters.py:45: [W0141(bad-builtin), GlanceImageFilterTestCase.apply_filter] Used builtin function 'filter'
************* Module tests.cloudferrylib.utils.test_filters
tests/cloudferrylib/utils/test_filters.py:24: [E1101(no-member), BaseCFFiltersTestCase.test_base_class_has_get_filters_method] Instance of 'BaseCFFiltersTestCase' has no 'assertTrue' member
tests/cloudferrylib/utils/test_filters.py:35: [E1101(no-member), FilterYamlTestCase.test_reads_config_on_first_use] Instance of 'FilterYamlTestCase' has no 'assertFalse' member
tests/cloudferrylib/utils/test_filters.py:38: [E1101(no-member), FilterYamlTestCase.test_reads_config_on_first_use] Instance of 'FilterYamlTestCase' has no 'assertTrue' member
tests/cloudferrylib/utils/test_filters.py:83: [E1101(no-member), FilterYamlTestCase.test_returns_instances_from_instance_ids] Instance of 'FilterYamlTestCase' has no 'assertTrue' member
tests/cloudferrylib/utils/test_filters.py:100: [E1101(no-member), FilterYamlTestCase.test_returns_images_from_image_ids] Instance of 'FilterYamlTestCase' has no 'assertTrue' member
tests/cloudferrylib/utils/test_filters.py:118: [E1101(no-member), FilterYamlTestCase.test_date_filter_returns_datetime_object] Instance of 'FilterYamlTestCase' has no 'assertTrue' member
************* Module cloudferrylib.os.image.glance_image
cloudferrylib/os/image/glance_image.py:157: [W0212(protected-access), GlanceImage.get_ref_image] Access to a protected member _resp of a client class
cloudferrylib/os/image/glance_image.py:238: [W0613(unused-argument), GlanceImage.read_info] Unused argument 'kwargs'
cloudferrylib/os/image/glance_image.py:279: [W1202(logging-format-interpolation), GlanceImage.make_image_info] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:290: [W1202(logging-format-interpolation), GlanceImage.make_image_info] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:293: [W1202(logging-format-interpolation), GlanceImage.make_image_info] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:303: [W0221(arguments-differ), GlanceImage.deploy] Arguments number differs from overridden 'deploy' method
cloudferrylib/os/image/glance_image.py:333: [W1202(logging-format-interpolation), GlanceImage.deploy] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:346: [W1202(logging-format-interpolation), GlanceImage.deploy] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:352: [W1202(logging-format-interpolation), GlanceImage.deploy] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:361: [W1202(logging-format-interpolation), GlanceImage.deploy] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:364: [W1202(logging-format-interpolation), GlanceImage.deploy] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:384: [W1202(logging-format-interpolation), GlanceImage.deploy] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:417: [W1202(logging-format-interpolation), GlanceImage.deploy] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:443: [W1202(logging-format-interpolation), GlanceImage.deploy] Use % formatting in logging functions but pass the % parameters as arguments

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by Settis** [comment]
retest this please

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module tests.cloudferrylib.os.image.test_image_filters
tests/cloudferrylib/os/image/test_image_filters.py:45: [W0141(bad-builtin), GlanceImageFilterTestCase.apply_filter] Used builtin function 'filter'
************* Module tests.cloudferrylib.utils.test_filters
tests/cloudferrylib/utils/test_filters.py:24: [E1101(no-member), BaseCFFiltersTestCase.test_base_class_has_get_filters_method] Instance of 'BaseCFFiltersTestCase' has no 'assertTrue' member
tests/cloudferrylib/utils/test_filters.py:35: [E1101(no-member), FilterYamlTestCase.test_reads_config_on_first_use] Instance of 'FilterYamlTestCase' has no 'assertFalse' member
tests/cloudferrylib/utils/test_filters.py:38: [E1101(no-member), FilterYamlTestCase.test_reads_config_on_first_use] Instance of 'FilterYamlTestCase' has no 'assertTrue' member
tests/cloudferrylib/utils/test_filters.py:83: [E1101(no-member), FilterYamlTestCase.test_returns_instances_from_instance_ids] Instance of 'FilterYamlTestCase' has no 'assertTrue' member
tests/cloudferrylib/utils/test_filters.py:100: [E1101(no-member), FilterYamlTestCase.test_returns_images_from_image_ids] Instance of 'FilterYamlTestCase' has no 'assertTrue' member
tests/cloudferrylib/utils/test_filters.py:118: [E1101(no-member), FilterYamlTestCase.test_date_filter_returns_datetime_object] Instance of 'FilterYamlTestCase' has no 'assertTrue' member
************* Module cloudferrylib.os.image.glance_image
cloudferrylib/os/image/glance_image.py:157: [W0212(protected-access), GlanceImage.get_ref_image] Access to a protected member _resp of a client class
cloudferrylib/os/image/glance_image.py:238: [W0613(unused-argument), GlanceImage.read_info] Unused argument 'kwargs'
cloudferrylib/os/image/glance_image.py:279: [W1202(logging-format-interpolation), GlanceImage.make_image_info] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:290: [W1202(logging-format-interpolation), GlanceImage.make_image_info] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:293: [W1202(logging-format-interpolation), GlanceImage.make_image_info] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:303: [W0221(arguments-differ), GlanceImage.deploy] Arguments number differs from overridden 'deploy' method
cloudferrylib/os/image/glance_image.py:333: [W1202(logging-format-interpolation), GlanceImage.deploy] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:346: [W1202(logging-format-interpolation), GlanceImage.deploy] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:352: [W1202(logging-format-interpolation), GlanceImage.deploy] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:361: [W1202(logging-format-interpolation), GlanceImage.deploy] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:364: [W1202(logging-format-interpolation), GlanceImage.deploy] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:384: [W1202(logging-format-interpolation), GlanceImage.deploy] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:417: [W1202(logging-format-interpolation), GlanceImage.deploy] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:443: [W1202(logging-format-interpolation), GlanceImage.deploy] Use % formatting in logging functions but pass the % parameters as arguments

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by Settis** [comment]
retest this please

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module tests.cloudferrylib.os.image.test_image_filters
tests/cloudferrylib/os/image/test_image_filters.py:45: [W0141(bad-builtin), GlanceImageFilterTestCase.apply_filter] Used builtin function 'filter'
************* Module tests.cloudferrylib.utils.test_filters
tests/cloudferrylib/utils/test_filters.py:24: [E1101(no-member), BaseCFFiltersTestCase.test_base_class_has_get_filters_method] Instance of 'BaseCFFiltersTestCase' has no 'assertTrue' member
tests/cloudferrylib/utils/test_filters.py:35: [E1101(no-member), FilterYamlTestCase.test_reads_config_on_first_use] Instance of 'FilterYamlTestCase' has no 'assertFalse' member
tests/cloudferrylib/utils/test_filters.py:38: [E1101(no-member), FilterYamlTestCase.test_reads_config_on_first_use] Instance of 'FilterYamlTestCase' has no 'assertTrue' member
tests/cloudferrylib/utils/test_filters.py:83: [E1101(no-member), FilterYamlTestCase.test_returns_instances_from_instance_ids] Instance of 'FilterYamlTestCase' has no 'assertTrue' member
tests/cloudferrylib/utils/test_filters.py:100: [E1101(no-member), FilterYamlTestCase.test_returns_images_from_image_ids] Instance of 'FilterYamlTestCase' has no 'assertTrue' member
tests/cloudferrylib/utils/test_filters.py:118: [E1101(no-member), FilterYamlTestCase.test_date_filter_returns_datetime_object] Instance of 'FilterYamlTestCase' has no 'assertTrue' member
************* Module cloudferrylib.os.image.glance_image
cloudferrylib/os/image/glance_image.py:157: [W0212(protected-access), GlanceImage.get_ref_image] Access to a protected member _resp of a client class
cloudferrylib/os/image/glance_image.py:238: [W0613(unused-argument), GlanceImage.read_info] Unused argument 'kwargs'
cloudferrylib/os/image/glance_image.py:279: [W1202(logging-format-interpolation), GlanceImage.make_image_info] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:290: [W1202(logging-format-interpolation), GlanceImage.make_image_info] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:293: [W1202(logging-format-interpolation), GlanceImage.make_image_info] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:303: [W0221(arguments-differ), GlanceImage.deploy] Arguments number differs from overridden 'deploy' method
cloudferrylib/os/image/glance_image.py:333: [W1202(logging-format-interpolation), GlanceImage.deploy] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:346: [W1202(logging-format-interpolation), GlanceImage.deploy] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:352: [W1202(logging-format-interpolation), GlanceImage.deploy] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:361: [W1202(logging-format-interpolation), GlanceImage.deploy] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:364: [W1202(logging-format-interpolation), GlanceImage.deploy] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:384: [W1202(logging-format-interpolation), GlanceImage.deploy] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:417: [W1202(logging-format-interpolation), GlanceImage.deploy] Use % formatting in logging functions but pass the % parameters as arguments
cloudferrylib/os/image/glance_image.py:443: [W1202(logging-format-interpolation), GlanceImage.deploy] Use % formatting in logging functions but pass the % parameters as arguments

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: tests/cloudferrylib/os/image/test_image_filters.py [hunk 1]
```diff
@@ -21,84 +22,103 @@
...
+        filter_yaml.get_tenant.return_value = tenant
+        filter_yaml.get_image_ids.return_value = image_ids
+        filter_yaml.get_image_date.return_value = date
 
-        glance_filters = filters.GlanceFilters(glance_client=mock.MagicMock(),
+        glance_filters = filters.GlanceFilters(glance_client=glance_client,
                                                filter_yaml=filter_yaml)
 
-        fs = glance_filters.get_filters()
+        filtered_images = filter(glance_filters.get_filter(), images)
```
Comment (plomakin) [inline tests/cloudferrylib/os/image/test_image_filters.py]: Is this logic correct? Are really you want to use builtin filter() here?

File: tests/cloudferrylib/os/image/test_image_filters.py [hunk 2]
```diff
@@ -21,84 +22,103 @@
...
 
-        for f in fs:
-            images = filter(f, images)
```
Comment (Settis) [inline tests/cloudferrylib/os/image/test_image_filters.py]: @plomakin builtin filter() was already here. I just keep it.

Comment (plomakin) [inline tests/cloudferrylib/os/image/test_image_filters.py]: k.

**Comment by plomakin** [comment]
+1 looks good to me.

File: cloudferrylib/utils/filters.py [hunk 3]
```diff
@@ -66,7 +66,7 @@ def __init__(self, filter_yaml):
         self.filter_yaml = filter_yaml
 
     @abc.abstractmethod
-    def get_filters(self):
+    def get_filter(self):
```
Comment (roman-verchikov) [inline cloudferrylib/utils/filters.py]: This change will break cinder volume functionality. Let's keep the interface and return list of filters, it seems to be more flexible.

Comment (Settis) [inline cloudferrylib/utils/filters.py]: Done

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by Settis** [comment]
retest this please

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by roman-verchikov** [comment]
+1

**Comment by plomakin** [comment]
+1, seems working fine ö

**Pull request merged**